src/app/components/core/asset-readings/readings-graph/readings-graph.component.ts

```diff
@@ -492,80 +492,95 @@ export class ReadingsGraphComponent implements OnDestroy {
-      } else {
-        return;
-      }
+      } 
+    } else {
+      assetFormatedReadings = this.readingFormate(readings);
+      this.timestamps = readings.reverse().map((r: any) => r.timestamp);
     }
-    this.timestamps = readings.reverse().map((r: any) => r.timestamp);
-    
+   
+    this.imageReadings = assetFormatedReadings.imageReadings.length > 0 ? this.getImage(assetFormatedReadings.imageReadings) : [];
+    this.numberTypeReadingsList = assetFormatedReadings.numReadings.length > 0 ? this.mergeObjects(assetFormatedReadings.numReadings) : [];
+    this.arrayTypeReadingsList = assetFormatedReadings.arrReadings.length > 0 ? this.mergeObjects(assetFormatedReadings.arrReadings) : [];
+    this.stringTypeReadingsList = mapValues(groupBy(assetFormatedReadings.strReadings,
+      (reading) => this.dateFormatter.transform(reading.timestamp, 'YYYY-MM-DD HH:mm:ss.SSS')), rlist => rlist.map(read => omit(read, 'timestamp')));
+
     this.setGraphStartTimestamp(optedTime);
     this.setInfoTextTimestamps();
     this.setGraphMinimumZoomValue();
+    this.setTabData(optedTime);
+  }
 
-    for (const r of readings) {
-      Object.entries(r.reading).forEach(([k, value]) => {
-        // discard unuseful reading
-        if (value === 'Data removed for brevity') {
-          return;
-        }
-        if (typeof value === 'number') {
-          numReadings.push({
-            key: k,
-            read: { x: r.timestamp, y: value }
-          });
-        } else if (typeof value === 'string') {
-          if (value.includes("__DPIMAGE")) {
-            this.getImageReadingsDimensions(value);
-            imageReadings.push({
-              datapoint: k,
-              imageData: value,
-              timestamp: this.dateFormatter.transform(r.timestamp, 'YYYY-MM-DD HH:mm:ss.SSS')
-            });
-          } else {
-            strReadings.push({
-              key: k,
-              timestamp: r.timestamp,
-              data: value
-            });
-          }
-        } else if (Array.isArray(value)) {
-          arrReadings.push({
-            key: k,
-            read: value,
-            timestamp: r.timestamp
+
+ readingFormate(readings) { 
+  const numReadings = [];
+  const strReadings = [];
+  const arrReadings = [];
+  const imageReadings = [];
+  for (const r of readings) {
+    Object.entries(r.reading).forEach(([k, value]) => {
+      // discard unuseful reading
+      if (value === 'Data removed for brevity') {
+        return;
+      }
+      if (typeof value === 'number') {
+        numReadings.push({
+          key:k,
+          read: { x: r.timestamp, y: value }
+        });
+      } else if (typeof value === 'string') {
+        if (value.includes("__DPIMAGE")) {
+          this.getImageReadingsDimensions(value);
+          imageReadings.push({
+            datapoint:  k,
+            imageData: value,
+            timestamp: this.dateFormatter.transform(r.timestamp, 'YYYY-MM-DD HH:mm:ss.SSS')
```

Comment:
We should probably define ts_fmt = 'YYYY-MM-DD HH:mm:ss.SSS' as const on top, as this is being used at multiples places